lesson5-LINQ/Task1/LinqTask.cs

```diff
@@ -8,85 +10,114 @@ public static class LinqTask
+
+            return result;
         }
 
         public static IEnumerable<Linq7CategoryGroup> Linq7(IEnumerable<Product> products)
         {
-            /* example of Linq7result
-
-             category - Beverages
-	            UnitsInStock - 39
-		            price - 18.0000
-		            price - 19.0000
-	            UnitsInStock - 17
-		            price - 18.0000
-		            price - 19.0000
-             */
-
-            throw new NotImplementedException();
+            var result = products.GroupBy(p => p.Category).Select(g => new Linq7CategoryGroup
+            {
+                Category = g.Key,
+                UnitsInStockGroup = g.GroupBy(c => c.UnitsInStock).Select(p => new Linq7UnitsInStockGroup
+                {
+                    Prices = g.Select(product => product.UnitPrice)
+                }).OrderBy(price => price)
+            });
+
+            return result;
         }
 
-        public static IEnumerable<(decimal category, IEnumerable<Product> products)> Linq8(
-            IEnumerable<Product> products,
-            decimal cheap,
-            decimal middle,
-            decimal expensive
-        )
+        public static IEnumerable<(decimal category, IEnumerable<Product> products)> Linq8(IEnumerable<Product> products, decimal cheap, decimal middle, decimal expensive)
         {
-            throw new NotImplementedException();
+            var result = products.GroupBy(p => new
+            {
+                isCheap = p.UnitPrice <= cheap,
+                isMiddle = p.UnitPrice <= middle & p.UnitPrice > cheap,
+                isExpensive = p.UnitPrice <= expensive & p.UnitPrice > middle
+            }).Select(g =>
+            {
+                if (g.Key.isExpensive)
+                    return (expensive, g.AsEnumerable());
+                return g.Key.isMiddle ? (middle, g.AsEnumerable()) : (cheap, g.AsEnumerable());
+            }).ToList();
+
+            return result;
         }
 
-        public static IEnumerable<(string city, int averageIncome, int averageIntensity)> Linq9(
-            IEnumerable<Customer> customers
-        )
+        public static IEnumerable<(string city, int averageIncome, int averageIntensity)> Linq9(IEnumerable<Customer> customers)
         {
-            throw new NotImplementedException();
+            var result = customers
+                .GroupBy(c => c.City)
+                .Select(g => 
+                    (g.Key, 
+                        (int)Math.Round(g.Average(c => c.Orders.Sum(o => o.Total))), 
+                        (int)Math.Round(g.Average(c => c.Orders.Length))));
+            
+            return result;
         }
 
         public static string Linq10(IEnumerable<Supplier> suppliers)
         {
-            throw new NotImplementedException();
+            var result = string.Join("", suppliers
+                .Select(s => s.Country)
+                .Distinct()
+                .OrderBy(c => c.Length)
+                .ThenBy(c => c[0]));
```

Comment:
possible exception if supplier doesn't have a country

Reply:
we can use `.ThenBy(c => c?[0])` here